libraries/bot-builder/src/main/java/com/microsoft/bot/builder/skills/SkillHandlerImpl.java

```diff
@@ -0,0 +1,252 @@
+import com.microsoft.bot.schema.ConversationReference;
+import com.microsoft.bot.schema.ResourceResponse;
+import org.apache.commons.lang3.NotImplementedException;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+/**
+ * This class inherited all the implementations of {@link SkillHandler} class since we needed similar code for {@link CloudSkillHandler}.
+ * The {@link CloudSkillHandler} class differs from {@link SkillHandler} class only in authentication by making use of {@link BotFrameworkAuthentication} class.
+ * This class is internal since it is only used in skill handler classes.
+ */
+public class SkillHandlerImpl {
+
+    private final String skillConversationReferenceKey;
+    private final BotAdapter adapter;
+    private final Bot bot;
+    private final SkillConversationIdFactoryBase conversationIdFactory;
+    private final Supplier<String> getOAuthScope;
+    private Logger logger = LoggerFactory.getLogger(SkillHandlerImpl.class);
+
+    public SkillHandlerImpl(
+        String withSkillConversationReferenceKey,
+        BotAdapter withAdapter,
+        Bot withBot,
+        SkillConversationIdFactoryBase withConversationIdFactory,
+        Supplier<String> withGetOAuthScope) {
+        if (StringUtils.isBlank(withSkillConversationReferenceKey)) {
+            throw new IllegalArgumentException("skillConversationReferenceKey cannot be null");
+        }
+        if (withAdapter == null) {
+            throw new IllegalArgumentException("adapter cannot be null");
+        }
+
+        if (withBot == null) {
+            throw new IllegalArgumentException("bot cannot be null");
+        }
+
+        this.skillConversationReferenceKey = withSkillConversationReferenceKey;
+        this.adapter = withAdapter;
+        this.bot = withBot;
+        this.conversationIdFactory = withConversationIdFactory;
+        this.getOAuthScope = withGetOAuthScope;
+    }
+
+    public CompletableFuture<ResourceResponse> onSendToConversation(
+        ClaimsIdentity claimsIdentity,
+        String conversationId,
+        Activity activity) {
+        return processActivity(claimsIdentity, conversationId, null, activity);
+    }
+
+    public CompletableFuture<ResourceResponse> onReplyToActivity(
+        ClaimsIdentity claimsIdentity,
+        String conversationId,
+        String activityId,
+        Activity activity) {
+        return processActivity(claimsIdentity, conversationId, activityId, activity);
+    }
+
+    public CompletableFuture<Void> onDeleteActivity(
+        ClaimsIdentity claimsIdentity,
+        String conversationId,
+        String activityId) {
+
+        SkillConversationReference skillConversationReference = getSkillConversationReference(conversationId).join();
```

Comment:
this is quite different from TS so I guess it's only based on C#. Is the join() a good idea?

Reply:
Agree with you, but it was a copy/paste from `SkillHandler`. As you mentioned it's only based on [C#](https://github.com/microsoft/botbuilder-dotnet/blob/main/libraries/Microsoft.Bot.Builder/Skills/SkillHandlerImpl.cs#L62-L73) apparently.

I don't think that the `join` is the best approach.